File: src/dere_daemon/routers/missions.py
"""Mission management endpoints."""

from __future__ import annotations

import asyncio
from datetime import UTC, datetime

from croniter import croniter
from fastapi import APIRouter, Depends, HTTPException, Request
from loguru import logger
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from dere_daemon.dependencies import get_db
from dere_daemon.missions.schedule_parser import is_valid_cron, parse_natural_language_schedule
from dere_shared.models import (
    Mission,
    MissionExecution,
    MissionStatus,
    MissionTriggerType,
)

router = APIRouter(prefix="/missions", tags=["missions"])


# Request/Response models
class CreateMissionRequest(BaseModel):
    """Request to create a new mission."""

    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    prompt: str = Field(..., min_length=1)
    schedule: str = Field(..., description="Cron expression or natural language schedule")
    personality: str | None = None
    allowed_tools: list[str] | None = None
    mcp_servers: list[str] | None = None
    plugins: list[str] | None = None
    thinking_budget: int | None = None
    model: str = "claude-sonnet-4-20250514"
    working_dir: str = "/workspace"
    sandbox_mode: bool = True
    sandbox_mount_type: str = "none"
    run_once: bool = False


class UpdateMissionRequest(BaseModel):
    """Request to update a mission."""

    name: str | None = None
    description: str | None = None
    prompt: str | None = None
    schedule: str | None = None
    personality: str | None = None
    allowed_tools: list[str] | None = None
    mcp_servers: list[str] | None = None
    plugins: list[str] | None = None
    thinking_budget: int | None = None
    model: str | None = None
    working_dir: str | None = None
    sandbox_mode: bool | None = None
    sandbox_mount_type: str | None = None
    run_once: bool | None = None


class MissionResponse(BaseModel):
    """Mission details response."""

    id: int
    name: str
    description: str | None
    prompt: str
    cron_expression: str
    natural_language_schedule: str | None
    timezone: str
    status: str
    next_execution_at: datetime | None
    last_execution_at: datetime | None
    personality: str | None
    allowed_tools: list[str] | None
    mcp_servers: list[str] | None
    plugins: list[str] | None
    thinking_budget: int | None
    model: str
    working_dir: str
    sandbox_mode: bool
    sandbox_mount_type: str
    run_once: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ExecutionResponse(BaseModel):
    """Mission execution details response."""

    id: int
    mission_id: int
    status: str
    trigger_type: str
    triggered_by: str | None
    started_at: datetime | None
    completed_at: datetime | None
    output_text: str | None
    output_summary: str | None
    tool_count: int
    error_message: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


async def _parse_schedule(schedule: str) -> tuple[str, str, str | None]:
    """Parse schedule input (cron or NL) into cron expression.

    Returns: (cron_expression, timezone, natural_language_schedule)
    """
    # Try as cron first
    if is_valid_cron(schedule):
        return schedule, "UTC", None

    # Parse as natural language
    cron_expr, timezone = await parse_natural_language_schedule(schedule)
    return cron_expr, timezone, schedule


@router.post("", response_model=MissionResponse)
async def create_mission(
    req: CreateMissionRequest,
    db: AsyncSession = Depends(get_db),
):
    """Create a new mission.

    Accepts either a cron expression or natural language schedule.
    """
    try:
        cron_expr, timezone, nl_schedule = await _parse_schedule(req.schedule)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e

    # Calculate first execution time
    now = datetime.now(UTC)
    next_run = croniter(cron_expr, now).get_next(datetime)

    mission = Mission(
        name=req.name,
        description=req.description,
        prompt=req.prompt,
        cron_expression=cron_expr,
        natural_language_schedule=nl_schedule,
        timezone=timezone,
        next_execution_at=next_run,
        personality=req.personality,
        allowed_tools=req.allowed_tools,
        mcp_servers=req.mcp_servers,
        plugins=req.plugins,
        thinking_budget=req.thinking_budget,
        model=req.model,
        working_dir=req.working_dir,
        sandbox_mode=req.sandbox_mode,
        sandbox_mount_type=req.sandbox_mount_type,
        run_once=req.run_once,
    )

    db.add(mission)
    await db.commit()
    await db.refresh(mission)

    logger.info("Created mission {}: {} (next run: {})", mission.id, mission.name, next_run)

    return mission


@router.get("", response_model=list[MissionResponse])
async def list_missions(
    db: AsyncSession = Depends(get_db),
    status: str | None = None,
):
    """List all missions, optionally filtered by status."""
    stmt = select(Mission).order_by(Mission.created_at.desc())

    if status:
        stmt = stmt.where(Mission.status == status)

    result = await db.execute(stmt)
    return list(result.scalars().all())


@router.get("/{mission_id}", response_model=MissionResponse)
async def get_mission(
    mission_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get mission details."""
    mission = await db.get(Mission, mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")
    return mission


@router.patch("/{mission_id}", response_model=MissionResponse)
async def update_mission(
    mission_id: int,
    req: UpdateMissionRequest,
    db: AsyncSession = Depends(get_db),
):
    """Update a mission."""
    mission = await db.get(Mission, mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")

    # Update fields if provided
    update_data = req.model_dump(exclude_unset=True)

    # Handle schedule change specially
    if "schedule" in update_data:
        schedule = update_data.pop("schedule")
        try:
            cron_expr, timezone, nl_schedule = await _parse_schedule(schedule)
            mission.cron_expression = cron_expr
            mission.timezone = timezone
            mission.natural_language_schedule = nl_schedule

            # Recalculate next execution
            now = datetime.now(UTC)
            mission.next_execution_at = croniter(cron_expr, now).get_next(datetime)
        except ValueError as e:
            raise HTTPException(status_code=400, detail=str(e)) from e

    # Apply other updates
    for field, value in update_data.items():
        setattr(mission, field, value)

    mission.updated_at = datetime.now(UTC)

    await db.commit()
    await db.refresh(mission)

    logger.info("Updated mission {}: {}", mission.id, mission.name)

    return mission


@router.delete("/{mission_id}")
async def delete_mission(
    mission_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Delete a mission and its execution history."""
    mission = await db.get(Mission, mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")

    await db.delete(mission)
    await db.commit()

    logger.info("Deleted mission {}: {}", mission_id, mission.name)

    return {"status": "deleted", "id": mission_id}


@router.post("/{mission_id}/pause", response_model=MissionResponse)
async def pause_mission(
    mission_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Pause a mission (prevents scheduled execution)."""
    mission = await db.get(Mission, mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")

    mission.status = MissionStatus.PAUSED.value
    mission.updated_at = datetime.now(UTC)

    await db.commit()
    await db.refresh(mission)

    logger.info("Paused mission {}: {}", mission.id, mission.name)

    return mission


@router.post("/{mission_id}/resume", response_model=MissionResponse)
async def resume_mission(
    mission_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Resume a paused mission."""
    mission = await db.get(Mission, mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")

    mission.status = MissionStatus.ACTIVE.value

    # Recalculate next execution from now
    now = datetime.now(UTC)
    mission.next_execution_at = croniter(mission.cron_expression, now).get_next(datetime)
    mission.updated_at = now

    await db.commit()
    await db.refresh(mission)

    logger.info("Resumed mission {}: {} (next run: {})", mission.id, mission.name, mission.next_execution_at)

    return mission


@router.post("/{mission_id}/execute")
async def trigger_execution(
    mission_id: int,
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    """Manually trigger a mission execution.

    The execution runs asynchronously; use GET /missions/{id}/executions to check status.
    """
    mission = await db.get(Mission, mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")

    # Get executor from app state
    executor = getattr(request.app.state, "mission_executor", None)
    if not executor:
        raise HTTPException(status_code=503, detail="Mission executor not available")

    # Execute asynchronously (don't block request)
    asyncio.create_task(
        executor.execute(
            mission,
            trigger_type=MissionTriggerType.MANUAL.value,
            triggered_by="user",
        )
    )

    logger.info("Manually triggered mission {}: {}", mission.id, mission.name)

    return {"status": "triggered", "mission_id": mission_id}


@router.get("/{mission_id}/executions", response_model=list[ExecutionResponse])
async def list_executions(
    mission_id: int,
    db: AsyncSession = Depends(get_db),
    limit: int = 50,
):
    """Get execution history for a mission."""
    # Verify mission exists
    mission = await db.get(Mission, mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")

    stmt = (
        select(MissionExecution)
        .where(MissionExecution.mission_id == mission_id)
        .order_by(MissionExecution.started_at.desc())
        .limit(limit)
    )
    result = await db.execute(stmt)
    return list(result.scalars().all())


@router.get("/{mission_id}/executions/{execution_id}", response_model=ExecutionResponse)
async def get_execution(
    mission_id: int,
    execution_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get details of a specific execution."""
    execution = await db.get(MissionExecution, execution_id)
    if not execution or execution.mission_id != mission_id:
        raise HTTPException(status_code=404, detail="Execution not found")
    return execution

File: packages/daemon/src/missions.ts
import type { Hono } from "hono";

import { getDb } from "./db.js";
import { getMissionExecutor } from "./mission-runtime.js";
import { getNextCronRun, parseSchedule } from "./mission-schedule.js";

const DEFAULT_MODEL = "claude-sonnet-4-20250514";

type MissionRow = {
  id: number;
  name: string;
  description: string | null;
  prompt: string;
  cron_expression: string;
  natural_language_schedule: string | null;
  timezone: string;
  status: string;
  next_execution_at: Date | null;
  last_execution_at: Date | null;
  personality: string | null;
  allowed_tools: string[] | null;
  mcp_servers: string[] | null;
  plugins: string[] | null;
  thinking_budget: number | null;
  model: string;
  working_dir: string;
  sandbox_mode: boolean;
  sandbox_mount_type: string;
  sandbox_settings: unknown;
  run_once: boolean;
  created_at: Date | null;
  updated_at: Date | null;
};

type MissionExecutionRow = {
  id: number;
  mission_id: number;
  status: string;
  trigger_type: string;
  triggered_by: string | null;
  started_at: Date | null;
  completed_at: Date | null;
  output_text: string | null;
  output_summary: string | null;
  tool_count: number | null;
  error_message: string | null;
  created_at: Date | null;
};

function nowDate(): Date {
  return new Date();
}

async function parseJson<T>(req: Request): Promise<T | null> {
  try {
    return (await req.json()) as T;
  } catch {
    return null;
  }
}

function toStringArray(value: unknown): string[] | null {
  if (!Array.isArray(value)) {
    return null;
  }
  const items = value.filter((item) => typeof item === "string") as string[];
  return items.length > 0 ? items : [];
}

function toMissionResponse(row: MissionRow): MissionRow {
  return row;
}

function toExecutionResponse(row: MissionExecutionRow): MissionExecutionRow {
  return row;
}

export function registerMissionRoutes(app: Hono): void {
  app.post("/missions", async (c) => {
    const payload = await parseJson<Record<string, unknown>>(c.req.raw);
    if (!payload) {
      return c.json({ error: "Invalid JSON payload" }, 400);
    }

    const name = payload.name;
    const prompt = payload.prompt;
    const schedule = payload.schedule;

    if (typeof name !== "string" || !name.trim()) {
      return c.json({ error: "name is required" }, 400);
    }
    if (typeof prompt !== "string" || !prompt.trim()) {
      return c.json({ error: "prompt is required" }, 400);
    }
    if (typeof schedule !== "string" || !schedule.trim()) {
      return c.json({ error: "schedule is required" }, 400);
    }

    let scheduleResult: {
      cron_expression: string;
      timezone: string;
      natural_language_schedule: string | null;
    };
    try {
      scheduleResult = await parseSchedule(schedule);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return c.json({ error: message }, 400);
    }

    let nextRun: Date;
    try {
      nextRun = getNextCronRun(scheduleResult.cron_expression, nowDate(), scheduleResult.timezone);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return c.json({ error: message }, 400);
    }

    const db = await getDb();
    const now = nowDate();
    const inserted = await db
      .insertInto("missions")
      .values({
        name: name.trim(),
        description: typeof payload.description === "string" ? payload.description : null,
        prompt: prompt.trim(),
        cron_expression: scheduleResult.cron_expression,
        natural_language_schedule: scheduleResult.natural_language_schedule,
        timezone: scheduleResult.timezone,
        next_execution_at: nextRun,
        personality: typeof payload.personality === "string" ? payload.personality : null,
        allowed_tools: toStringArray(payload.allowed_tools),
        mcp_servers: toStringArray(payload.mcp_servers),
        plugins: toStringArray(payload.plugins),
        thinking_budget:
          typeof payload.thinking_budget === "number" ? payload.thinking_budget : null,
        model: typeof payload.model === "string" ? payload.model : DEFAULT_MODEL,
        working_dir: typeof payload.working_dir === "string" ? payload.working_dir : "/workspace",
        sandbox_mode: typeof payload.sandbox_mode === "boolean" ? payload.sandbox_mode : true,
        sandbox_mount_type:
          typeof payload.sandbox_mount_type === "string" ? payload.sandbox_mount_type : "none",
        sandbox_settings:
          payload.sandbox_settings && typeof payload.sandbox_settings === "object"
            ? (payload.sandbox_settings as Record<string, unknown>)
            : null,
        run_once: typeof payload.run_once === "boolean" ? payload.run_once : false,
        status: "active",
        last_execution_at: null,
        user_id: null,
        created_at: now,
        updated_at: now,
      })
      .returningAll()
      .executeTakeFirstOrThrow();

    console.log(
      `[missions] created mission ${inserted.id} (${inserted.name}) next=${nextRun.toISOString()}`,
    );

    return c.json(toMissionResponse(inserted));
  });

  app.get("/missions", async (c) => {
    const status = c.req.query("status");
    const db = await getDb();
    let query = db.selectFrom("missions").selectAll().orderBy("created_at", "desc");
    if (status) {
      query = query.where("status", "=", status);
    }
    const rows = await query.execute();
    return c.json(rows.map((row) => toMissionResponse(row)));
  });

  app.get("/missions/:mission_id", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    if (!Number.isFinite(missionId)) {
      return c.json({ error: "Invalid mission_id" }, 400);
    }

    const db = await getDb();
    const mission = await db
      .selectFrom("missions")
      .selectAll()
      .where("id", "=", missionId)
      .executeTakeFirst();

    if (!mission) {
      return c.json({ error: "Mission not found" }, 404);
    }

    return c.json(toMissionResponse(mission));
  });

  app.patch("/missions/:mission_id", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    if (!Number.isFinite(missionId)) {
      return c.json({ error: "Invalid mission_id" }, 400);
    }

    const payload = await parseJson<Record<string, unknown>>(c.req.raw);
    if (!payload) {
      return c.json({ error: "Invalid JSON payload" }, 400);
    }

    const db = await getDb();
    const mission = await db
      .selectFrom("missions")
      .selectAll()
      .where("id", "=", missionId)
      .executeTakeFirst();

    if (!mission) {
      return c.json({ error: "Mission not found" }, 404);
    }

    const updates: Record<string, unknown> = { updated_at: nowDate() };

    if (typeof payload.name === "string") {
      updates.name = payload.name;
    }
    if (typeof payload.description === "string" || payload.description === null) {
      updates.description = payload.description as string | null;
    }
    if (typeof payload.prompt === "string") {
      updates.prompt = payload.prompt;
    }
    if (typeof payload.personality === "string" || payload.personality === null) {
      updates.personality = payload.personality as string | null;
    }
    if ("allowed_tools" in payload) {
      updates.allowed_tools = toStringArray(payload.allowed_tools);
    }
    if ("mcp_servers" in payload) {
      updates.mcp_servers = toStringArray(payload.mcp_servers);
    }
    if ("plugins" in payload) {
      updates.plugins = toStringArray(payload.plugins);
    }
    if (typeof payload.thinking_budget === "number" || payload.thinking_budget === null) {
      updates.thinking_budget = payload.thinking_budget as number | null;
    }
    if (typeof payload.model === "string") {
      updates.model = payload.model;
    }
    if (typeof payload.working_dir === "string") {
      updates.working_dir = payload.working_dir;
    }
    if (typeof payload.sandbox_mode === "boolean") {
      updates.sandbox_mode = payload.sandbox_mode;
    }
    if (typeof payload.sandbox_mount_type === "string") {
      updates.sandbox_mount_type = payload.sandbox_mount_type;
    }
    if (payload.sandbox_settings && typeof payload.sandbox_settings === "object") {
      updates.sandbox_settings = payload.sandbox_settings as Record<string, unknown>;
    }
    if (typeof payload.run_once === "boolean") {
      updates.run_once = payload.run_once;
    }

    if (typeof payload.schedule === "string" && payload.schedule.trim()) {
      try {
        const scheduleResult = await parseSchedule(payload.schedule);
        const nextRun = getNextCronRun(
          scheduleResult.cron_expression,
          nowDate(),
          scheduleResult.timezone,
        );
        updates.cron_expression = scheduleResult.cron_expression;
        updates.timezone = scheduleResult.timezone;
        updates.natural_language_schedule = scheduleResult.natural_language_schedule;
        updates.next_execution_at = nextRun;
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        return c.json({ error: message }, 400);
      }
    }

    const updated = await db
      .updateTable("missions")
      .set(updates)
      .where("id", "=", missionId)
      .returningAll()
      .executeTakeFirstOrThrow();

    return c.json(toMissionResponse(updated));
  });

  app.delete("/missions/:mission_id", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    if (!Number.isFinite(missionId)) {
      return c.json({ error: "Invalid mission_id" }, 400);
    }

    const db = await getDb();
    const mission = await db
      .selectFrom("missions")
      .select(["id", "name"])
      .where("id", "=", missionId)
      .executeTakeFirst();

    if (!mission) {
      return c.json({ error: "Mission not found" }, 404);
    }

    await db.deleteFrom("missions").where("id", "=", missionId).execute();

    console.log(`[missions] deleted mission ${missionId} (${mission.name})`);
    return c.json({ status: "deleted", id: missionId });
  });

  app.post("/missions/:mission_id/pause", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    if (!Number.isFinite(missionId)) {
      return c.json({ error: "Invalid mission_id" }, 400);
    }

    const db = await getDb();
    const updated = await db
      .updateTable("missions")
      .set({ status: "paused", updated_at: nowDate() })
      .where("id", "=", missionId)
      .returningAll()
      .executeTakeFirst();

    if (!updated) {
      return c.json({ error: "Mission not found" }, 404);
    }

    return c.json(toMissionResponse(updated));
  });

  app.post("/missions/:mission_id/resume", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    if (!Number.isFinite(missionId)) {
      return c.json({ error: "Invalid mission_id" }, 400);
    }

    const db = await getDb();
    const mission = await db
      .selectFrom("missions")
      .selectAll()
      .where("id", "=", missionId)
      .executeTakeFirst();

    if (!mission) {
      return c.json({ error: "Mission not found" }, 404);
    }

    let nextRun: Date | null = null;
    try {
      nextRun = getNextCronRun(mission.cron_expression, nowDate(), mission.timezone);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      return c.json({ error: message }, 400);
    }

    const updated = await db
      .updateTable("missions")
      .set({
        status: "active",
        next_execution_at: nextRun,
        updated_at: nowDate(),
      })
      .where("id", "=", missionId)
      .returningAll()
      .executeTakeFirstOrThrow();

    return c.json(toMissionResponse(updated));
  });

  app.post("/missions/:mission_id/execute", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    if (!Number.isFinite(missionId)) {
      return c.json({ error: "Invalid mission_id" }, 400);
    }

    const db = await getDb();
    const mission = await db
      .selectFrom("missions")
      .selectAll()
      .where("id", "=", missionId)
      .executeTakeFirst();

    if (!mission) {
      return c.json({ error: "Mission not found" }, 404);
    }

    const executor = getMissionExecutor();
    if (!executor) {
      return c.json({ error: "Mission executor not available" }, 503);
    }

    void executor.execute(mission, "manual", "user").catch((error) => {
      console.log(`[missions] manual execution failed for ${missionId}: ${String(error)}`);
    });

    return c.json({ status: "triggered", mission_id: missionId });
  });

  app.get("/missions/:mission_id/executions", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    if (!Number.isFinite(missionId)) {
      return c.json({ error: "Invalid mission_id" }, 400);
    }

    const limitParam = c.req.query("limit");
    const limit = limitParam ? Math.max(1, Number(limitParam)) : 50;

    const db = await getDb();
    const mission = await db
      .selectFrom("missions")
      .select(["id"])
      .where("id", "=", missionId)
      .executeTakeFirst();

    if (!mission) {
      return c.json({ error: "Mission not found" }, 404);
    }

    const executions = await db
      .selectFrom("mission_executions")
      .selectAll()
      .where("mission_id", "=", missionId)
      .orderBy("started_at", "desc")
      .limit(Number.isFinite(limit) ? limit : 50)
      .execute();

    return c.json(executions.map((exec) => toExecutionResponse(exec)));
  });

  app.get("/missions/:mission_id/executions/:execution_id", async (c) => {
    const missionId = Number(c.req.param("mission_id"));
    const executionId = Number(c.req.param("execution_id"));
    if (!Number.isFinite(missionId) || !Number.isFinite(executionId)) {
      return c.json({ error: "Invalid mission_id or execution_id" }, 400);
    }

    const db = await getDb();
    const execution = await db
      .selectFrom("mission_executions")
      .selectAll()
      .where("id", "=", executionId)
      .executeTakeFirst();

    if (!execution || execution.mission_id !== missionId) {
      return c.json({ error: "Execution not found" }, 404);
    }

    return c.json(toExecutionResponse(execution));
  });
}
